File: xor/c/xor.c
#include "xor.h"

void xor_encrypt(char *message,
                 char *key,
                 char *output,
                 int msg_len,
                 int key_len)
{
    for(int i = 0; i < msg_len; i++)
    {
        output[i] = message[i] ^ key[i % key_len];
    }
}

void xor_decrypt(char *ciphertext,
                 char *key,
                 char *output,
                 int cipher_len,
                 int key_len)
{
    for(int i = 0; i < cipher_len; i++)
    {
        output[i] = ciphertext[i] ^ key[i % key_len];
    }
}

File: xor/ocaml/xor.ml
let xor_encrypt message key =
  let msg_len = Bytes.length message in
  let key_len = Bytes.length key in
  let output = Bytes.create msg_len in

  for i = 0 to msg_len - 1 do
    let m = Char.code (Bytes.get message i) in
    let k = Char.code (Bytes.get key (i mod key_len)) in
    Bytes.set output i (Char.chr (m lxor k))
  done;

  output

let xor_decrypt ciphertext key =
  let cipher_len = Bytes.length ciphertext in
  let key_len = Bytes.length key in
  let output = Bytes.create cipher_len in

  for i = 0 to cipher_len - 1 do
    let c = Char.code (Bytes.get ciphertext i) in
    let k = Char.code (Bytes.get key (i mod key_len)) in
    Bytes.set output i (Char.chr (c lxor k))
  done;

  output